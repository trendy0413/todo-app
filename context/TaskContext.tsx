'use client';

import React, { createContext, useReducer, useContext } from 'react';

export interface Task {
  id: string;
  title: string;
  description: string;
  status: 'Pending' | 'In Progress' | 'Completed';
  dueDate: string;
}

type TaskState = Task[];

interface TaskContextProps {
  tasks: TaskState;
  addTask: (task: Task) => void;
  editTask: (id: string, updatedTask: Task) => void;
  deleteTask: (id: string) => void;
}

const TaskContext = createContext<TaskContextProps | null>(null);

const taskReducer = (state: TaskState, action: any): TaskState => {
  switch (action.type) {
    case 'ADD_TASK':
      return [...state, action.payload];
    case 'EDIT_TASK':
      return state.map((task) =>
        task.id === action.payload.id ? action.payload : task
      );
    case 'DELETE_TASK':
      return state.filter((task) => task.id !== action.payload);
    default:
      return state;
  }
};

export const TaskProvider = ({ children }: { children: React.ReactNode }) => {
  const [tasks, dispatch] = useReducer(taskReducer, []);

  const addTask = (task: Task) => dispatch({ type: 'ADD_TASK', payload: task });
  const editTask = (id: string, updatedTask: Task) =>
    dispatch({ type: 'EDIT_TASK', payload: { id, updatedTask } });
  const deleteTask = (id: string) => dispatch({ type: 'DELETE_TASK', payload: id });

  return (
    <TaskContext.Provider value={{ tasks, addTask, editTask, deleteTask }}>
      {children}
    </TaskContext.Provider>
  );
};

export const useTasks = () => {
  const context = useContext(TaskContext);
  if (!context) throw new Error('useTasks must be used within a TaskProvider');
  return context;
};
