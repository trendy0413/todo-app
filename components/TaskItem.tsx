'use client';

import React from 'react';
import { Task } from '@/context/TaskContext';

interface TaskItemProps {
  task: Task;
}

const TaskItem: React.FC<TaskItemProps> = React.memo(({ task }) => {
  const isOverdue = new Date(task.dueDate) < new Date();

  return (
    <div
      className={`${
        isOverdue ? 'bg-red-50' : 'bg-gray-50'
      } border border-gray-200 rounded-lg p-4 transition-shadow hover:shadow-md`}
      data-testid="task-item"
    >
      <h2 className="font-bold text-lg">{task.title}</h2>
      <p className="text-gray-700">{task.description}</p>
      <p className="text-sm text-gray-500">Due: {task.dueDate}</p>
      <p
        className={`text-sm ${
          task.status === 'Completed' ? 'text-green-600' : 'text-red-600'
        }`}
      >
        {task.status}
      </p>
    </div>
  );
});

TaskItem.displayName = 'TaskItem';

export default TaskItem;
