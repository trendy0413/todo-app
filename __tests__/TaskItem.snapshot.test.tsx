import renderer from 'react-test-renderer';
import TaskItem from '@/components/TaskItem';
import { Task } from '@/context/TaskContext';

describe('TaskItem Snapshots', () => {
  it('renders TaskItem snapshot correctly', () => {
    const task: Task = {
      id: '1',
      title: 'Snapshot Task',
      description: 'Snapshot Description',
      status: 'Pending',
      dueDate: '2024-12-31',
    };

    const tree = renderer.create(<TaskItem task={task} />).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('renders overdue TaskItem snapshot correctly', () => {
    const overdueTask: Task = {
      id: '2',
      title: 'Overdue Task',
      description: 'Past due date',
      status: 'Pending',
      dueDate: '2023-01-01',
    };

    const tree = renderer.create(<TaskItem task={overdueTask} />).toJSON();
    expect(tree).toMatchSnapshot();
  });
});
