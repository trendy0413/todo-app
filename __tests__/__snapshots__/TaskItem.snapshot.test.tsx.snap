// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TaskItem Snapshots renders TaskItem snapshot correctly 1`] = `
<div
  className="bg-gray-50 border border-gray-200 rounded-lg p-4 transition-shadow hover:shadow-md"
  data-testid="task-item"
>
  <h2
    className="font-bold text-lg"
  >
    Snapshot Task
  </h2>
  <p
    className="text-gray-700"
  >
    Snapshot Description
  </p>
  <p
    className="text-sm text-gray-500"
  >
    Due: 
    2024-12-31
  </p>
  <p
    className="text-sm text-red-600"
  >
    Pending
  </p>
</div>
`;

exports[`TaskItem Snapshots renders overdue TaskItem snapshot correctly 1`] = `
<div
  className="bg-red-50 border border-gray-200 rounded-lg p-4 transition-shadow hover:shadow-md"
  data-testid="task-item"
>
  <h2
    className="font-bold text-lg"
  >
    Overdue Task
  </h2>
  <p
    className="text-gray-700"
  >
    Past due date
  </p>
  <p
    className="text-sm text-gray-500"
  >
    Due: 
    2023-01-01
  </p>
  <p
    className="text-sm text-red-600"
  >
    Pending
  </p>
</div>
`;

exports[`renders TaskItem snapshot correctly 1`] = `
<div
  className="bg-gray-50 border border-gray-200 rounded-lg p-4 transition-shadow hover:shadow-md"
>
  <h2
    className="font-bold text-lg"
  >
    Snapshot Task
  </h2>
  <p
    className="text-gray-700"
  >
    Snapshot Description
  </p>
  <p
    className="text-sm text-gray-500"
  >
    Due: 
    2024-12-31
  </p>
  <p
    className="text-sm text-red-600"
  >
    Pending
  </p>
</div>
`;
